### DISCUSSION TITLE
INDEX NOT LOADING FROM DISK

### DISCUSSION THREAD

**Discussion by Sha-x2-nk** [facebookresearch/faiss#3565] [General]
# Summary

I created a faiss index and train it as follows:
faiss::IndexHNSWFlat* generate_index(int m = 128, int ef_search = 250, int ef_construction = 64){
	faiss::IndexHNSWFlat* index = new faiss::IndexHNSWFlat(512, m, faiss::METRIC_INNER_PRODUCT);
	index->hnsw.efSearch = ef_search;
	index->hnsw.efConstruction = ef_construction;

	return index;
}
faiss::write_index(index, faiss_index_path.c_str());
std::cout<<"SAVED INDEX TO "<<faiss_index_path<<std::endl;

Now when I try to load the index and use it as 
faiss::IndexHNSWFlat* index;
faiss::Index *loaded = faiss::read_index(faiss_index_path.c_str());
index = static_cast<faiss::IndexHNSWFlat*>(loaded);
std::vector<float> embed_vec(embed.begin<float>(), embed.end<float>());
std::vector<long int> labels(5);
std::vector<float> distances(5);
index->search(1, embed_vec.data(), 5, distances.data(), labels.data());
where embed_vec is a 512 dimensional embedding, i get the following output
pure virtual method called
terminate called without an active exception
Aborted (core dumped)

Installed from: built from source.

Faiss compilation options:using mkl

Running on:
- CPU
Interface: 
- C++

**Comment by mdouze** [comment]
please don't use `static_cast`, always use `dynamic_cast`.

**Comment by Sha-x2-nk** [comment]
hey there. I also tried with dynamic_cast and it is still not working. 
right now I am using index->search only without casting it to HNSWFlat index
